internal/eventstream/client.go

```diff
@@ -0,0 +1,415 @@
+		group.Go(func() error {
+			err := client.checkMissedChanges(groupCtx, objType)
+			if err != nil {
+				client.Logger.Errorw("Replaying API events resulted in errors",
+					zap.String("object type", objType), zap.Error(err))
+			}
+			return err
+		})
+	}
+
+	finCh := make(chan struct{})
+	go func() {
+		err := group.Wait()
+		if err != nil {
+			client.Logger.Errorw("Replaying the API resulted in errors", zap.Error(err), zap.Duration("duration", time.Since(startTime)))
+		} else {
+			client.Logger.Infow("Replaying the API has finished", zap.Duration("duration", time.Since(startTime)))
+		}
+
+		close(finCh)
+	}()
+
+	return finCh
+}
+
+// worker is the Client's main background worker, taking care of event.Event dispatching and mode switching.
+//
+// When the Client is in the replay phase, requested by replayPhaseRequest, events from the Event Stream API will
+// be cached until the replay phase has finished, while replayed events will be delivered directly.
+//
+// Communication takes place over the channels eventDispatcherEventStream, eventDispatcherReplay and replayPhaseRequest.
+func (client *Client) worker() {
+	// eventCache is a subset of event.Event used for caching those in replayCache below.
+	type eventCache struct {
+		SourceId int64
+		Name     string
+		Type     string
+	}
+
+	var (
+		// replayFinCh holds a reference when the Client is in the replay phase. It will be closed from the producer,
+		// startReplayWorker, when replaying is finished, which indicates the select below to switch phases. When this
+		// variable is nil, the Client is in the normal operating phase.
+		replayFinCh chan struct{}
+
+		// replayBuffer holds Event Stream events to be replayed after the replay phase has finished.
+		replayBuffer = make([]*event.Event, 0)
+		// replayCache maps eventCache derived from event.Events to API time to skip replaying outdated events.
+		replayCache = make(map[eventCache]time.Time)
+	)
+
+	// replayCacheUpdate updates the replayCache if this eventMsg seems to be the latest of its kind.
+	replayCacheUpdate := func(ev *eventMsg) {
+		evc := eventCache{ev.event.SourceId, ev.event.Name, ev.event.Type}
+		ts, ok := replayCache[evc]
+		if !ok || ev.apiTime.After(ts) {
+			replayCache[evc] = ev.apiTime
+		}
+	}
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing down main worker as context is finished", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.replayPhaseRequest:
+			if replayFinCh != nil {
+				// There shouldn't be multiple concurrent startReplayWorker calls. However, technically this is possible,
+				// i.e. when the Icinga 2 API connection is flapping.
+				client.Logger.Error("Another replay phase request was sent while the Client is already replaying")
+				break
+			}
+
+			client.Logger.Debug("Dispatcher enters replay phase, starting caching Event Stream events")
+			replayFinCh = client.startReplayWorker()
```

Comment:
The idea of the replay (or maybe better named catch up phase?) is to gather events than have been lost during an event stream disconnect. Currently, it looks like this is implemented so that this catching up only happens on connection loss, if the previous catch up was completed.

To get the events as reliably as we can, the requests to the objects API must be performed after the event stream API responded (i.e. started streaming any new events that occur from then on). Otherwise something like the following sequence would lose events:
1. Event stream connects.
2. Objects API is queried and all the requests are completed successfully.
3. Now imagine the notifications daemon has some kind of weird network issue while it's still processing the events generated based on the objects API where the event stream connection is reset but the database connection just hangs for a bit.
4. Event stream reconnects but doesn't start a new replay phase because the previous one didn't complete yet.

Any events between the last one read from the old event stream and the first one from the new event stream would be lost now.